lib/browser/tag/each.js

```diff
@@ -170,79 +132,76 @@ export default function _each(dom, parent, expr) {
-        if (i == tags.length) {
-          if (isVirtual)
-            makeVirtual(tag, frag)
-          else frag.appendChild(domToInsert)
-        }
-        // this tag must be insert
-        else {
-          if (isVirtual)
-            makeVirtual(tag, root, tags[i])
-          else root.insertBefore(domToInsert, tags[i].root)
-          oldItems.splice(i, 0, item)
-        }
 
-        tags.splice(i, 0, tag)
-        if (child) arrayishAdd(parent.tags, tagName, tag, true)
-        pos = i // handled here so no move
-      } else tag.update(item)
-
-      // reorder the tag if it's not located in its previous position
-      if (pos !== i && _mustReorder) {
-        // update the DOM
-        if (isVirtual)
-          moveVirtual(tag, root, tags[i], dom.childNodes.length)
-        else if (tags[i].root.parentNode) root.insertBefore(tag.root, tags[i].root)
-        // update the position attribute if it exists
-        if (expr.pos)
-          tag[expr.pos] = i
-        // move the old tag instance
-        tags.splice(i, 0, tags.splice(pos, 1)[0])
-        // move the old item
-        oldItems.splice(i, 0, oldItems.splice(pos, 1)[0])
-        // if the loop tags are not custom
-        // we need to move all their custom tags into the right position
-        if (!child && tag.tags) moveNestedTags(tag, i)
+        if (child)
+          arrayishAdd(parent.tags, tagName, tag, true)
+
+        // cache parent tag internally
+        defineProperty(tag, '_parent', parent)
+        // this will be used in events bound to this node
+        tag._item = item
+      } else {
+        tag.update(item)
+
+        if (!child && tag.tags)
+        // non-custom tags should have their nested tags moved
+          moveNestedTags(tag, i)
       }
 
-      // cache the original item to use it in the events bound to this node
-      // and its children
-      tag._item = item
-      // cache the real parent tag internally
-      defineProperty(tag, '_parent', parent)
+      tagsUpdated.push(tag)
 
+      tagsPending[tag._riot_id] = true
+
+      if (skipTo !== i) {
+        isVirtual ?
+          makeVirtual(tag, frag) :
+          frag.appendChild(tag.root)
+      }
     })
 
-    // remove the redundant tags
-    unmountRedundant(items, tags, tagName, parent)
-
-    // insert the new nodes
-    root.insertBefore(frag, ref)
-    if (isOption) {
-
-      // #1374 FireFox bug in <option selected={expression}>
-      if (FIREFOX && !root.multiple) {
-        for (var n = 0; n < root.length; n++) {
-          if (root[n].__riot1374) {
-            root.selectedIndex = n  // clear other options
-            delete root[n].__riot1374
-            break
-          }
+    let target = ref
+
+    // insert new nodes
+    if (skipTo >= 0) {
+      let tag = tagsPrevious[skipTo]
+
+      target = !tag._virts ?
+        tag.root.nextSibling :
+        tag._virts[tag._virts.length - 1].nextSibling
+    }
+
+    root.insertBefore(frag, target)
```

Comment:
I guess we can do all the DOM manipulations in a separate batch to avoid too many reflows